### <center> Dev Bharatbhai Patel(8938546) </center>
### <center>Lab-4</center>

In [72]:
# Importing necessary packages

import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score, mean_absolute_error, mean_absolute_percentage_error

* 1. Get the data, and run a train-validation-test split. Description of each column can be found in sklearn documentation. Look at the documentation for the load_diabetes method to know what are as_frame and scaled arguments are for.

In [75]:
# Loadin the diabetes datasets
diabetes_X, diabetes_y = datasets.load_diabetes(as_frame=True, scaled=True, return_X_y=True)

# Splitting the data into training, testing and Validation sets
X_train, X_test, Y_train, Y_test = train_test_split(diabetes_X, diabetes_y, test_size=0.2, random_state=2)
X_train, x_val, Y_train, y_val = train_test_split(X_train, Y_train, test_size=0.2, random_state=2)

* 2. Run a multivariate linear regression on all variables

In [76]:
# Multivariate Linear Regression

Linear_Regression = LinearRegression()
Linear_Regression.fit(X_train, Y_train)

LinearRegression()

In [77]:
# Predicting the target values
predict_1 = Linear_Regression.predict(X_test)

# Printing thr Coefficients and Intercept
print('Coefficients: ', Linear_Regression.coef_)
print('Intercept: ', Linear_Regression.intercept_)

Coefficients:  [  -73.75657995  -266.40794388   517.94194727   370.66497105
 -1251.98617314   891.75390971   323.78686007   112.25743252
  1054.48855519    30.62504975]
Intercept:  150.07377278777557


* 3. Run a polynomial regression of the 2nd degree on the BMI feature alone

In [78]:
# Reshaping and extracting the variables to BMI

X_train_BMI = np.array(X_train["bmi"]).reshape(-1, 1)
X_test_BMI = np.array(X_test["bmi"]).reshape(-1, 1)
X_val_BMI = np.array(x_val["bmi"]).reshape(-1, 1)

# Training the models by polynomial regression with 2nd degree
ply_regression = PolynomialFeatures(degree=2, include_bias=False)

X_train_ply_regression = ply_regression.fit_transform(X_train_BMI)
X_test_ply_regression = ply_regression.fit_transform(X_test_BMI)
X_val_ply_regression = ply_regression.fit_transform(X_val_BMI)

polynomial_reression = LinearRegression()
polynomial_reression.fit(X_train_ply_regression, Y_train)


LinearRegression()

In [79]:
# Predicting the target values
predict_2 = polynomial_reression.predict(X_test_ply_regression)

# Printing thr Coefficients and Intercept
print('Coefficients: ', polynomial_reression.coef_)
print('Intercept: ', polynomial_reression.intercept_)

Coefficients:  [939.50489618 248.87395634]
Intercept:  151.58063663264556


* 4. Run a multivariate polynomial regression of the 2nd degree on all variables (Hint: set include_bias=False in PolynomialFeatures) 

In [80]:
# Training the models by multivariate polynomial regression with 2nd degree
Multivariate_ply_regression = PolynomialFeatures(degree=2, include_bias=False)

X_train_Mlt_ply_regression = Multivariate_ply_regression.fit_transform(X_train)
X_test_Mlt_ply_regression = Multivariate_ply_regression.fit_transform(X_test)
X_val_Mlt_ply_regression = Multivariate_ply_regression.fit_transform(x_val)


Mlt_ply_regression = LinearRegression()
Mlt_ply_regression.fit(X_train_Mlt_ply_regression, Y_train)

LinearRegression()

In [81]:
# Predicting the target value
predict_3 = Mlt_ply_regression.predict(X_test_Mlt_ply_regression)

# Printing thr Coefficients and Intercept
print('Coefficients: ', Mlt_ply_regression.coef_)
print('Intercept: ', Mlt_ply_regression.intercept_)

Coefficients:  [ 6.25235497e+00 -2.73674376e+02  4.06894814e+02  3.63490828e+02
 -6.84752957e+03  5.78567630e+03  2.46736103e+03  1.16557104e+02
  3.01458546e+03  1.07778042e+02  1.76125281e+03  4.51789881e+03
  4.90248165e+02 -2.33541087e+03 -5.70312468e+03 -3.95254147e+02
  6.93044767e+03  7.59697758e+03  4.07445060e+01  1.42746436e+03
 -1.65257782e+00  1.36528560e+03  1.60080421e+03  2.35762549e+04
 -2.23734165e+04 -4.48577422e+03  3.44082331e+03 -1.04018284e+04
  2.01338315e+03  1.45056734e+03  1.13374742e+03  1.12775812e+04
 -7.66671946e+03 -7.93065273e+03 -9.03968781e+03 -2.68461226e+03
  3.37158708e+03  1.14672217e+03  8.66576524e+03 -5.43280465e+03
 -3.41242958e+03 -2.78807927e+03  5.80298276e+02 -3.49894331e+03
  1.36475766e+05 -1.99143780e+05 -1.22561333e+05 -3.94376542e+04
 -1.09690526e+05 -1.76081800e+04  7.46015781e+04  9.12039342e+04
  2.08758171e+04  8.44436838e+04  1.41841897e+04  2.61390743e+04
  1.56981948e+04  5.00449099e+04  1.04905548e+04  7.19463338e+03
  1.531536

* 5. Compare the three models by looking at R-squared, MAPE and MAE. Explain what the values mean for a non-expert and add your insight about the values of each model. Note: You can add any further comparisons and code (this is not necessary for a perfect score, but will be reviewed and evaluated)

In [96]:
# Creating a function for Comparing all the models
def Models_Comp(y_t,y_pre):
    print(f'MAPE: {(mean_absolute_percentage_error(y_t, y_pre)*100):.2f}%')
    print(f'MAE: {mean_absolute_error(y_t, y_pre):.2f}')
    print(f'Performance {r2_score(y_t,y_pre)*100:.2f}% ')

# Comparing the data on training models
print('Multi Linear Regression')
Models_Comp(Y_test,predict_1)

print('\nPolynomial Regession model on BMI')
Models_Comp(Y_test,predict_2)

print('\nPolynomial Regession model on All data')
Models_Comp(Y_test,predict_3)


Multi Linear Regression
MAPE: 35.91%
MAE: 45.17
Performance 42.63% 

Polynomial Regession model on BMI
MAPE: 42.98%
MAE: 49.42
Performance 29.76% 

Polynomial Regession model on All data
MAPE: 35.46%
MAE: 48.01
Performance 34.80% 


In [105]:
# Comparing the data on testing models
print('Multi Linear Regression')
Linear_Regression_test = Linear_Regression.predict(X_train)
Models_Comp(Y_train,Linear_Regression_test)

polynomial_reression_test = polynomial_reression.predict(X_test_ply_regression)
print('\nPolynomial Regession model on BMI')
Models_Comp(Y_test,polynomial_reression_test)

Mlt_ply_regression_test = Mlt_ply_regression.predict(X_test_Mlt_ply_regression)
print('\nPolynomial Regession model on All data')
Models_Comp(Y_test,Mlt_ply_regression_test)

Multi Linear Regression
MAPE: 39.14%
MAE: 43.95
Performance 52.73% 

Polynomial Regession model on BMI
MAPE: 42.98%
MAE: 49.42
Performance 29.76% 

Polynomial Regession model on All data
MAPE: 35.46%
MAE: 48.01
Performance 34.80% 


In [106]:
# Comparing the data on validating models
print('Multi Linear Regression')
Linear_Regression_test = Linear_Regression.predict(x_val)
Models_Comp(y_val,Linear_Regression_test)

polynomial_reression_test = polynomial_reression.predict(X_val_ply_regression)
print('\nPolynomial Regession model on BMI')
Models_Comp(y_val,polynomial_reression_test)

Mlt_ply_regression_test = Mlt_ply_regression.predict(X_val_Mlt_ply_regression)
print('\nPolynomial Regession model on All data')
Models_Comp(y_val,Mlt_ply_regression_test)

Multi Linear Regression
MAPE: 40.27%
MAE: 40.16
Performance 51.70% 

Polynomial Regession model on BMI
MAPE: 46.83%
MAE: 45.19
Performance 44.06% 

Polynomial Regession model on All data
MAPE: 40.80%
MAE: 41.19
Performance 49.70% 


* From the above comparison of all the models `Multi Linear Regression` performance is better than other models.

- 6. Please answer the following questions
    -  How many parameters are we fitting for each of the three models? Explain these values. Hint: for explaining the parameters of the polynomial regression, you can use poly.get_feature_names_out()

    - 1. Multivariate linear regression: 
        -  **11** parameters, which inlcudes **Number of Features(10)** + **Intercept Term(1)** 

In [107]:
len(Linear_Regression.coef_) + 1

11

* Polynomial Regression of 2nd Degree on BMI:
    * **3** parameters, which includes **Number of Features(2)** + **Intercept Term(1)**

In [108]:
len(polynomial_reression.coef_) + 1

3

* Multivariate Polynomial Regression of the 2nd degree on all variables:
    * **66** parameters, which includes **Number of Features(65)** + **Intercept Term(1)**

In [109]:
len(Mlt_ply_regression.coef_) + 1

66

* 2. Which model would you choose for deployment, and why?
    *   Well, I will go for **Multivariate Linear Regression Model** because it has the lowest MAE with `43.95`, which shows that it has the accurate mean.
    *   It also has the highest performance rate with `52.73`.
    *   So, I will choose Multivariate Linear Regression Model because of the high performance rate and with low mean absolute error.